File: java/src/main/java/com/saucelabs/simplesauce/SauceSession.java
package com.saucelabs.simplesauce;

import lombok.Getter;
import org.openqa.selenium.JavascriptExecutor;
import lombok.Setter;
import org.openqa.selenium.InvalidArgumentException;
import org.openqa.selenium.MutableCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.net.MalformedURLException;
import java.net.URL;

public class SauceSession {

    /**
     * Defaults for username and access key are environment variables
     */
    private String username = System.getenv("SAUCE_USERNAME");
    private String accessKey = System.getenv("SAUCE_ACCESS_KEY");


    @Getter @Setter private DataCenter dataCenter = DataCenter.US_WEST;
    @Getter private final SauceOptions sauceOptions;
    @Setter private URL sauceUrl;

    @Getter private RemoteWebDriver driver;

    public SauceSession() {
        this(new SauceOptions());
    }

    public SauceSession(SauceOptions options) {
        sauceOptions = options;
    }

    public RemoteWebDriver start() {
        driver = createRemoteWebDriver(getSauceUrl(), sauceOptions.toCapabilities());
        return driver;
	}

    public URL getSauceUrl() {
        if (sauceUrl != null) {
            return sauceUrl;
        } else {
            String url = "https://" + getUsername() + ":" + getAccessKey() + "@" + dataCenter + "/wd/hub";
            try {
                return new URL(url);
            } catch (MalformedURLException e) {
                throw new InvalidArgumentException("Invalid URL");
            }
        }
    }

    protected RemoteWebDriver createRemoteWebDriver(URL url, MutableCapabilities capabilities) {
        return new RemoteWebDriver(url, capabilities);
    }

    protected JavascriptExecutor getJSExecutor() {
        return driver;
    }

    public void stop(Boolean passed) {
        String result = passed ? "passed" : "failed";
        stop(result);
    }

    public void stop(String result) {
        updateResult(result);
        stop();
    }

    private void updateResult(String result) {
        getJSExecutor().executeScript("sauce:job-result=" + result);
    }

    private void stop() {
        if(driver !=null) {
            driver.quit();
        }
    }

    private String getUsername() {
        if (username != null) {
            return username;
        } else if (System.getProperty("SAUCE_USERNAME") != null) {
            return System.getProperty("SAUCE_USERNAME");
        } else {
            throw new SauceEnvironmentVariablesNotSetException("Sauce Username was not provided");
        }
    }

    private String getAccessKey() {
        if (accessKey != null) {
            return accessKey;
        } else if (System.getProperty("SAUCE_ACCESS_KEY") != null) {
            return System.getProperty("SAUCE_ACCESS_KEY");
        } else {
            throw new SauceEnvironmentVariablesNotSetException("Sauce Access Key was not provided");
        }
    }
}

File: dotnet/Simple.Sauce/SauceSession.cs
﻿using OpenQA.Selenium;
using System;
using System.Collections.Generic;

namespace Sauce.Bindings
{
    public class SauceSession
    {
        private Uri _sauceUrl;

        public SauceSession()
        {
            Driver = new SauceDriver();
            Options = new SauceOptions();
        }

        public SauceSession(ISauceRemoteDriver driver)
        {
            Driver = driver;
            Options = new SauceOptions();
        }

        public SauceSession(SauceOptions options)
        {
            Options = options;
            Driver = new SauceDriver();
        }

        public SauceSession(SauceOptions options, ISauceRemoteDriver driver)
        {
            Options = options;
            Driver = driver;
        }

        public DataCenter DataCenter { get; set; } = DataCenter.UsWest;
        public SauceOptions Options { get; }

        public ISauceRemoteDriver Driver { get; set; }

        public IWebDriver Start()
        {
            if (Options.BrowserName.Value == Browser.Firefox.Value)
                return CreateFirefoxDriver();
            if (!string.IsNullOrEmpty(Options.ConfiguredSafariOptions.BrowserVersion))
                return CreateSafariDriver();
            if (!string.IsNullOrEmpty(Options.ConfiguredFirefoxOptions.BrowserVersion))
                return CreateFirefoxDriver();
            return CreateChromeDriver();
        }

        private IWebDriver CreateFirefoxDriver()
        {
            var sauceUserName = Environment.GetEnvironmentVariable("SAUCE_USERNAME");
            var sauceAccessKey = Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");
            var sauceConfiguration = new Dictionary<string, object>
            {
                ["username"] = sauceUserName,
                ["accessKey"] = sauceAccessKey
            };

            Options.ConfiguredFirefoxOptions.AddAdditionalOption("sauce:options", sauceConfiguration);
            return Driver.CreateRemoteWebDriver(Options.ConfiguredFirefoxOptions);
        }

        private IWebDriver CreateSafariDriver()
        {
            var sauceUserName = Environment.GetEnvironmentVariable("SAUCE_USERNAME");
            var sauceAccessKey = Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");
            var sauceConfiguration = new Dictionary<string, object>
            {
                ["username"] = sauceUserName,
                ["accessKey"] = sauceAccessKey
            };

            Options.ConfiguredSafariOptions.AddAdditionalOption("sauce:options", sauceConfiguration);
            return Driver.CreateRemoteWebDriver(Options.ConfiguredSafariOptions);
        }

        private IWebDriver CreateChromeDriver()
        {
            var sauceUserName = Environment.GetEnvironmentVariable("SAUCE_USERNAME");
            var sauceAccessKey = Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");
            var sauceConfiguration = new Dictionary<string, object>
            {
                ["username"] = sauceUserName,
                ["accessKey"] = sauceAccessKey
            };

            Options.ConfiguredChromeOptions.AddAdditionalOption("sauce:options", sauceConfiguration);
            return Driver.CreateRemoteWebDriver(Options.ConfiguredChromeOptions);
        }

        private IWebDriver CreateEdgeBrowser()
        {
            var sauceUserName = Environment.GetEnvironmentVariable("SAUCE_USERNAME");
            var sauceAccessKey = Environment.GetEnvironmentVariable("SAUCE_ACCESS_KEY");
            var sauceConfiguration = new Dictionary<string, object>
            {
                ["username"] = sauceUserName,
                ["accessKey"] = sauceAccessKey
            };

            Options.ConfiguredEdgeOptions.AddAdditionalOption("sauce:options", sauceConfiguration);
            return Driver.CreateRemoteWebDriver(Options.ConfiguredEdgeOptions);
        }

        public void Stop(bool isPassed)
        {
            if (Driver is null)
                return;
            var script = "sauce:job-result=" + (isPassed ? "passed" : "failed");
            Driver.ExecuteScript(script);
            Driver.Quit();
        }

        public Uri GetSauceUrl()
        {
            return _sauceUrl ?? new Uri(DataCenter.Value);
        }
    }
}